### PR TITLE
Implement asset properties widget

### PR DESCRIPTION
**Pull request by ghost**
This cumulocity plugin implements Asset properties widget

### CODE REVIEW

File: README.md
Comment (srinath0707) [inline README.md:1]: Add a complete readme. Full description of this widget functionality, if possible, with some images.

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts
Comment (srinath0707) [inline asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts:1]: Can we revisit this later to use asset-properties-item component form sub-assets module.

File: asset-properties-widget/common/preview.ts
Comment (srinath0707) [inline asset-properties-widget/common/preview.ts:1]: Why is this required?

File: debug.log
Comment (srinath0707) [inline debug.log:1]: delete this

File: gulpfile.js
Comment (srinath0707) [inline gulpfile.js:1]: why is this required? can be deleted?

File: package.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    "@angular/compiler": "14.0.6",
+    "@angular/core": "14.0.6",
+    "@angular/forms": "14.0.6",
+    "@angular/material": "^14.2.7",
+    "@angular/platform-browser": "14.0.6",
+    "@angular/platform-browser-dynamic": "14.0.6",
+    "@angular/router": "14.0.6",
+    "@angular/upgrade": "14.0.6",
+    "@c8y/client": "1016.0.85",
+    "@c8y/ngx-components": "1016.0.85",
```
Comment (srinath0707) [inline package.json]: why not use the latest c8y versions?

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts
Comment (amar-c8y) [reply asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts:1]: yes agreed will reuse the sub-assets module later.

File: asset-properties-widget/common/preview.ts
Comment (amar-c8y) [reply asset-properties-widget/common/preview.ts:1]: this file used to show constant image in widget preview page.

File: debug.log
Comment (amar-c8y) [reply debug.log:1]: Incorporated

File: gulpfile.js
Comment (amar-c8y) [reply gulpfile.js:1]: this file is used to create local build and create zip file and also if dist file is exist then first it will delete and create a new zip file.

File: package.json (hunk 1 above)
Comment (amar-c8y) [reply package.json]: upgrade to 1018 latest version.

File: README.md
Comment (amar-c8y) [reply README.md:1]: yes, I will add README file.

File: asset-properties-widget/common/asset-property-constant.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export const defaultProperty = [
+{ c8y_JsonSchema:{properties :{name:{type: "string",label: "Name"}} }, name: 'name', label: "Name", type: "string", active: true, isEditable:true, isExistingProperty:true},
```
Comment (spau6) [inline asset-properties-widget/common/asset-property-constant.ts:2]: Won't isEditable should be false?

Comment (amar-c8y) [reply asset-properties-widget/common/asset-property-constant.ts:2]: we are allowed to edit name here as in DTM support to edit asset name.

File: locales/de.po
Comment (spau6) [inline locales/de.po:1]: I cannot see the translation for this widget, add the strings to the PO file

File: asset-properties-widget/component/asset-properties-config/asset-property-selector.component.css
Comment (spau6) [inline asset-properties-widget/component/asset-properties-config/asset-property-selector.component.css:1]: Is this file required?

File: asset-properties-widget/component/asset-properties-config/asset-property-selector.component.html [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  </thead>
+
+  <tbody class="header-row" cdkDropList (cdkDropListDropped)="drop($event)">
+    <tr cdkDrag cdkDragLockAxis="y" class="property" *ngFor="let property of properties" class="table-row-not-expanded">
+      <td cdkDragHandle style="width: 5%" *ngIf="!property.isHide">
+        <i c8yicon="bars" class="dlt-c8y-icon-bars" style="cursor: move"></i>
+      </td>
+      <td style="width: 10%" *ngIf="!property.isHide">
+        <input type="checkbox" (change)="updateOptions()" [(ngModel)]="property.active" />
+        <!-- <a *ngIf="isComplexProperty(property)"
```
Comment (spau6) [inline asset-properties-widget/component/asset-properties-config/asset-property-selector.component.html]: Is this required? remove commented code

File: asset-properties-widget/component/asset-properties-config/asset-property-selector.component.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        this.assetPropertySelectorModalRef.hide();
+      });
+    }
+    constructCustomProperties(properties:IManagedObject[]){
+      this.properties.forEach(property => {
+        if(this.isComplexProperty(property)){
+          let customProperties = property.c8y_JsonSchema.properties[property.c8y_JsonSchema.key]?.properties;
+          const tempCustomProperties = {};
+          for (const key in customProperties) {
+            if(customProperties[key].active == true){
```
Comment (spau6) [inline asset-properties-widget/component/asset-properties-config/asset-property-selector.component.ts]: make use of ===

File: asset-properties-widget/component/asset-properties-view/asset-properties-view.component.html [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<c8y-asset-properties
+  *ngIf = "!isEmptyWidget"
+  [properties]="properties"
+  [asset]="selectedAsset"
+  (assetChange)="assetChange($event)"
+  class="d-contents"
+></c8y-asset-properties>
+<div *ngIf = "isEmptyWidget" style="text-align: center">
+  No Data
```
Comment (spau6) [inline asset-properties-widget/component/asset-properties-view/asset-properties-view.component.html]: add translate pipe, also add in other html where you have missed

File: package.json [hunk 6]
```diff
@@ -0,0 +1,101 @@
+{
+  "name": "asset-properties-widget-plugin",
+  "version": "1.0.1",
+  "description": "Asset properties",
+  "author": "Amarjyoti Raj - Software AG, IoT Accelerators R&D",
```
Comment (spau6) [inline package.json]: We belong to IOT-Apps

File: yarn.lock
Comment (spau6) [inline yarn.lock:1]: Hope all lock files are updated ones

File: asset-properties-widget/component/asset-properties-config/asset-property-selector.component.css
Comment (amar-c8y) [reply asset-properties-widget/component/asset-properties-config/asset-property-selector.component.css:1]: deleted

File: asset-properties-widget/component/asset-properties-config/asset-property-selector.component.html (hunk 3 above)
Comment (amar-c8y) [reply asset-properties-widget/component/asset-properties-config/asset-property-selector.component.html]: removed

File: asset-properties-widget/component/asset-properties-config/asset-property-selector.component.ts (hunk 4 above)
Comment (amar-c8y) [reply asset-properties-widget/component/asset-properties-config/asset-property-selector.component.ts]: added

File: asset-properties-widget/component/asset-properties-view/asset-properties-view.component.html (hunk 5 above)
Comment (amar-c8y) [reply asset-properties-widget/component/asset-properties-view/asset-properties-view.component.html]: translation pipe added

File: yarn.lock
Comment (amar-c8y) [reply yarn.lock:1]: yes

File: package.json (hunk 6 above)
Comment (amar-c8y) [reply package.json]: yes, changed to IOT-Apps.

File: locales/de.po
Comment (amar-c8y) [reply locales/de.po:1]: will add in next iteration

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.html [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    <ng-container *ngIf="!isLoading">
+      <div
+        class="card m-b-8"
+        *ngFor="let prop of customProperties"
+        [ngClass]="{ 'card-highlight': prop.isEdit }"
+      >
+        <div class="card-block " [ngClass]="{ 'p-b-0': prop.isEdit }">
+          <div class="d-flex p-b-8" *ngIf="!prop.isEdit">
+            <p title="{{ prop.label | translate | humanize }}" class="text-medium text-truncate">
+              {{ prop?.label | translate | humanize }}
```
Comment (spau6) [inline asset-properties-widget/asset-properties-widget-view/asset-properties.component.html]: I dont think both translate and humanize will not work... refer core code

File: asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.html [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  </ng-container>
+  <ng-container *ngSwitchCase="'object'">
+    <ul class="list-unstyled c8y-custom-properties">
+      <li
+        *ngFor="let prop of complex; let i = index"
+        [ngClass]="{ 'separator-top-bottom': i === 0, 'separator-bottom': i > 0 }"
+        class="p-t-4 p-b-4 flex-row text-nowrap"
+      >
+        <label class="small m-b-0 m-r-8" [ngClass]="{ 'a-s-start': prop.file }">{{
+          prop.label | translate | humanize
```
Comment (spau6) [inline asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.html]: same as below comment

File: asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  }
+
+  private async resolveFile() {
+    if (this.file) {
+      try {
+        const imageFile = await this.filesService.getFile(this.file);
+        this.previewImage = await this.getPreviewIfImage(imageFile);
+      } catch (ex) {
+        this.alert.danger(gettext('File could not be loaded.'));
```
Comment (spau6) [inline asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.ts:62]: If you adding translation then make sure this should get covered

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    for (const key of keys) {
+      if (Array.isArray(model[key]) && model[key][0]?.file instanceof File) {
+        try {
+          const upload = await this.inventoryBinary.create(model[key][0].file);
+          try {
+            if (moId && moId[0]) {
+              await this.inventory.childAdditionsRemove(moId[0], this.asset.id);
+            }
+          } catch (ex) {
+            this.alert.addServerFailure(ex);
```
Comment (spau6) [inline asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts]: Incorporate https://github.softwareag.com/IOTA/cumulocity-ui/pull/4379 changes here

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        this.toggleEdit(prop);
+        return;
+      }
+      const updatedAsset = { id: this.asset.id, ...propertyValue };
+      const { data } = await this.inventory.update(updatedAsset);
+      this.toggleEdit(prop);
+      this.asset = data;
+      await this.loadAsset();
+      this.alert.success(gettext('Properties changes saved.'));
+    } catch (ex) {
```
Comment (spau6) [inline asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts:186]: Incorporate https://github.softwareag.com/IOTA/cumulocity-ui/pull/4379 changes here

File: asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.html (hunk 8 above)
Comment (amar-c8y) [reply asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.html]: incorporate

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.html (hunk 7 above)
Comment (amar-c8y) [reply asset-properties-widget/asset-properties-widget-view/asset-properties.component.html]: incorporate

File: asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.ts (hunk 9 above)
Comment (amar-c8y) [reply asset-properties-widget/asset-properties-widget-view/asset-properties-item.component.ts:62]: added to the PO file

File: asset-properties-widget/component/asset-property-selector-model/schema-property-selector.component.html [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    <button
+      class="btn btn-default"
+      type="button"
+      title="{{ 'Cancel' | translate }}"
+      (click)="onCancelButtonClicked()"
+    >
+      {{ 'Cancel' | translate }}
+    </button>
+    <button
+      class="btn btn-primary"
```
Comment (spau6) [inline asset-properties-widget/component/asset-property-selector-model/schema-property-selector.component.html:106]: add data-cy so that it will be easy for automation, please incorporate in all places of html

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts (hunk 11 above)
Comment (amar-c8y) [reply asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts:186]: incorporate

File: asset-properties-widget/component/asset-property-selector-model/schema-property-selector.component.html (hunk 12 above)
Comment (amar-c8y) [reply asset-properties-widget/component/asset-property-selector-model/schema-property-selector.component.html:106]: added data-cy

File: asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts (hunk 10 above)
Comment (amar-c8y) [reply asset-properties-widget/asset-properties-widget-view/asset-properties.component.ts]: incorporate

**Review by spau6 — APPROVED** [review]

**Review by amar-c8y — APPROVED** [review]
